bls/signature.go

```diff
@@ -41,15 +43,30 @@ func AggregateSignatures(signatures []*Signature) (*Signature, error) {
 	return &Signature{flowSignature: aggregateFlowSignature}, nil
 }
 
-// VerifyAggregateSignature takes in a slice of bls.PublicKeys, a bls.Signature, and a payload and returns
+// VerifyAggregateSignatureSinglePayload takes in a slice of bls.PublicKeys, a bls.Signature, and a single payload and returns
 // true if every bls.PublicKey in the slice signed the payload. The input bls.Signature is the aggregate
 // signature of each of their respective bls.Signatures for that payload.
-func VerifyAggregateSignature(publicKeys []*PublicKey, signature *Signature, payloadBytes []byte) (bool, error) {
-	var flowPublicKeys []flowCrypto.PublicKey
-	for _, publicKey := range publicKeys {
-		flowPublicKeys = append(flowPublicKeys, publicKey.flowPublicKey)
+func VerifyAggregateSignatureSinglePayload(publicKeys []*PublicKey, signature *Signature, payloadBytes []byte) (bool, error) {
+	flowPublicKeys := extractFlowPublicKeys(publicKeys)
+	return flowCrypto.VerifyBLSSignatureOneMessage(flowPublicKeys, signature.flowSignature, payloadBytes, hashingAlgorithm)
+}
+
+// VerifyAggregateSignatureMultiplePayloads takes in a slice of bls.PublicKeys, a bls.Signature, and a slice of payloads.
+// It returns true if each bls.PublicKey at index i has signed its respective payload at index i in the payloads slice.
+// The input bls.Signature is the aggregate signature of each public key's partial bls.Signatures for its respective payload.
+func VerifyAggregateSignatureMultiplePayloads(publicKeys []*PublicKey, signature *Signature, payloadsBytes [][]byte) (bool, error) {
```

Comment:
Hmm... I had a question about this. When I imagined it, it seemed like the signature would take a [][]*PublicKey to parallel the payloadBytes. But here I see it's only a single list of public keys.

If this is the case, then how would verification of the QCs work? My understanding was we'd have a situation like the following:
* HighQC1 is signed by {pk1, pk2, pk3}
* HighQC2 is signed by {pk4, pk5}
* HighQC3 is signed by {pk6, pk7, pk8}

In which case I would expect something more like the following:
publicKeys := [][]*PublicKey{
    {pk1, pk2, pk3},
    {pk4, pk5}
    {pk6, pk7, pk8}
}
payloads := [][]byte{
    HighQC1,
    HighQC2,
    HighQC3
}
verified, err := VerifyAggregateSignatureMultiplePayloads(publicKeys, signature, payloads)

I think I'm missing something though. Lmk what you think.

Reply:
I see. Is the intention here compression of the payloads? Or is there a limitation in BLS signature verification that requires payload deduplication?

My understanding of all of this is purely based off the UBC researchers' reference implementation here: https://github.com/deso-protocol/hotstuff_pseudocode/blob/diamondhands/REV-02-Jun-06-2023/fast_hotstuff_bls.go#L186. 

So I may be missing something too. 

My understanding is that, whenever those public keys timeout, it doesn't sign its high QC, but rather its high QC's view. 

Using your example with pk1, pk2, pk3... pk8, when each public key times out for view N, it signs the following:
- pk1 signs the pair <N, HighQC1.View>
- pk2 signs the pair <N, HighQC1.View>
- pk3 signs the pair <N, HighQC2.View>
- ...
- pk8 signs the pair <N, HighQC3.View>

When the leader aggregates a timeout QC for view N, it constructs the timeout QC as follows:
- publicKeys = [pk1, pk2, pk3, pk4, pk5, pk6, pk7, pk8]
  - all of those public keys have timed out
- payloads = [HighQC1.View, HighQC1.View, HighQC1.View, HighQC2.View, HighQC2.View, HighQC3.View, HighQC3.View, HighQC3.View]
  - the index of each element in this slice corresponds to the public key that sent it
- signature = aggregation of all of the partial signatures from pk1, pk2, pk3,...

Reply:
OK I get what you're doing now. Your understanding of what we need to do is correct, I was just imagining a different schema that "deduplicates" the payload, if that makes sense.

Basically, what you're saying is you aggregate a bunch of (PublicKey, payload) pairs, which I think is the most elegant thing to do. The less elegant thing that I was suggesting is we aggregate ([]PublicKey, payload) pairs, where each tuple will have a *unique* payload plus all the public keys that signed that particular unique payload. This amounts to the same thing, just grouped differently. Lmk if this doesn't make sense.

I think the way you have it is how we should keep it. It's technically more redundant because we repeat the same view number a lot in ValidatorsTimeoutHighQCViews but we can compress it easily later if it ever becomes an issue (e.g. when we have 10k+ validators or something...).

Reply:
Correct. And you're right that the payloads are duplicated in my current scheme. I'm actually aligned with you that deduplicating should be the end-game of what we want to implement.

What's interesting is that, even if we do the deduplication as you suggest, we'll still need to unravel the public keys and payloads into individual (publickey, payload) pairs to do the BLS signature verification. This means that the in-memory `TimeoutAggregateQuorumCertificate` struct's schema is best left as-is.

This also means that the biggest value of deduplicating the payloads is the compression it provides when sending timeout QCs as network messages. 

What nice about the above is that, it means we can implement the compression in the ToBytes/FromBytes functions for the timeout message without changing the struct fields at all. When we get to the point where we want to compress the views, we can increment the `MsgDeSoHeader` version and implement new byte encoding that hs the compression.

Reply:
Just read through this and I love the idea of pushing the compression into the ToBytes/FromBytes. That actually seems like the perfect place to put it because its benefit is to the network layer, no to the actual logic like you mention.